### PR TITLE
Load from mothur

### LINKED ISSUE #88: Loading Mothur data
We seem to miss Mothur importer, this could be added similar to [mia::loadFromQIIME2](https://github.com/microbiome/mia/blob/master/R/loadFromQIIME2.R).

We have one in [microbiome::read_mothur2phyloseq](https://github.com/microbiome/microbiome/blob/master/R/read_mothur2phyloseq.R) and this might help to get it started.

### PR DESCRIPTION
**Pull request by TuomasBorman**
Hi,

here is the implementation of `loadFromMothur`. I used loadFromQIIME2 and `microbiome::read_mothur2phykoseq` as references. 

I have tested this with the data (Baxter_FITs_Microbiome_2016) that I found from one of the older versions of `microbiome.` However, it did not include phylotree nor refseq, so I haven't tested those functionalities. 

With those abundance table, taxa data, and sample meta data, `loadFromMothur` seems to work. 

Issue: https://github.com/microbiome/mia/issues/88

-Tuomas

### CODE REVIEW

File: R/loadFromMothur.R
Comment (antagomir) [inline R/loadFromMothur.R]: Are the indentations here and below correct? Shouldn't it always be multiples of 4 by Bioc standards?

**Review by antagomir — COMMENTED** [review]
Seems good! Can be merged after my comments, if @FelixErnst agrees.

File: R/loadFromMothur.R
Comment (FelixErnst) [inline R/loadFromMothur.R]: Description text is missing. Please add it, even though it might be only one sentence

File: R/loadFromMothur.R
Comment (FelixErnst) [inline R/loadFromMothur.R]: I am not really familiar with mothur output. Are those data types available from a typical mothur run?

If not, please remove them for clarity.

File: R/loadFromMothur.R
Comment (FelixErnst) [inline R/loadFromMothur.R]: Adjust if necessary based on comment above

File: R/loadFromMothur.R
Comment (FelixErnst) [inline R/loadFromMothur.R]: wrong link

File: R/loadFromMothur.R
Comment (FelixErnst) [inline R/loadFromMothur.R]: Hard coded files are a no-go. Please add them as example data and reference them in the examples via `system.file`. Then the data can actually be loaded and the `\dontrun` section can be "removed" aka. the code actually be run

File: R/loadFromMothur.R
Comment (FelixErnst) [reply R/loadFromMothur.R]: yep those are not correct.

edit: the indentation

File: R/loadFromMothur.R
Comment (FelixErnst) [inline R/loadFromMothur.R]: This is not applicable here, since the `.read_qza` function is `QIIME2` output specific. You probably just need to add another function as warpper for `ape::read.tree`

File: R/loadFromMothur.R
Comment (FelixErnst) [inline R/loadFromMothur.R]: See comment above. This requires your own kind of wrapper, since the assumption for `.read_qza` are not assumable for mothur output.

File: R/loadFromMothur.R
Comment (FelixErnst) [inline R/loadFromMothur.R]: Why? They are supposed to end up in the `rowData`

File: R/loadFromMothur.R
Comment (FelixErnst) [inline R/loadFromMothur.R]: Where does `x` come from?

File: R/loadFromMothur.R
Comment (FelixErnst) [inline R/loadFromMothur.R:218]: Please remove those empty lines. (All across the PR)

File: R/loadFromMothur.R
Comment (FelixErnst) [inline R/loadFromMothur.R]: This section is a bit weird in the sense that the data stored as `rownames` here must match the `rownames` of the assay, if they are set. From the code this is not clearly managed which is the actual true feature name in the end returned by `rownames(tse)`.

File: R/loadFromMothur.R
Comment (FelixErnst) [inline R/loadFromMothur.R]: `rowData` will be a `DataFrame` in the end, so this conversion does make sense

File: R/loadFromMothur.R
Comment (FelixErnst) [inline R/loadFromMothur.R]: Why? Additional data is allowed isn't it? Why not import it as well?

And also, why is the subset necessary? Shouldn't the dimension of the data match out of the box?

File: R/loadFromMothur.R
Comment (FelixErnst) [inline R/loadFromMothur.R]: Same as above. Why don't the dimension match? Or the order?

I don't think that this hand-holding is necessary. If mothur supplies data which matches, we don't need to make sure, the user didn't jumble the data around.

If mothur supplies jumbled-up data, then I see the point.

**Review by FelixErnst — CHANGES_REQUESTED** [review]
Hi Tuoams,

thanks for the PR.

Some thing need to be done and some need to be discuss. Have a look the individual comments and let me know, what you think about the issues raised. Thanks!

Felix

**Comment by antagomir** [comment]
Great, I agree with all. Somehow many of these had escaped my attention, need to try and be more sharp next time!

File: R/loadFromMothur.R
Comment (TuomasBorman) [reply R/loadFromMothur.R]: Fixed

File: R/loadFromMothur.R
Comment (TuomasBorman) [reply R/loadFromMothur.R]: Added

File: R/loadFromMothur.R
Comment (TuomasBorman) [reply R/loadFromMothur.R]: I'm not familiar either with Mothur, so that is why I basically tried to just implemented the microbiome's function

It seems that Mothur output files can include:

- assay: https://mothur.org/wiki/shared_file/
- rowData: https://mothur.org/wiki/taxonomy_file/ or https://mothur.org/wiki/constaxonomy_file/
- colData: https://mothur.org/wiki/design_file/

(Full file type list: https://mothur.org/wiki/tags/#file_types)

So, these file types above should be allowed as input?
I should probably try to find more presentative data so I don't have to do this blindfolded

File: R/loadFromMothur.R
Comment (TuomasBorman) [reply R/loadFromMothur.R:218]: Ok, should I avoid empty rows also in the future? Personally, I find it easier to read code when it is little bit grouped, but for others it might be confusing, unnecessary, or annoying

File: R/loadFromMothur.R
Comment (FelixErnst) [reply R/loadFromMothur.R]: > So, these file types above should be allowed as input?

I guess so

File: R/loadFromMothur.R
Comment (FelixErnst) [reply R/loadFromMothur.R:218]: That is a fine line. Grouping steps is helpful, but a break using `#` with a short description would work for that as well. An empty line after or before a return is not helpful. Have a look at the other code in `mia`. I think you will get it.

File: R/loadFromMothur.R
Comment (microsud) [reply R/loadFromMothur.R]: Yes these are standard outputs of mother.

File: R/loadFromMothur.R
Comment (TuomasBorman) [reply R/loadFromMothur.R]: According to <https://mothur.org/wiki/shared_file/, 

- " label" is just first value of original data's first column. It is just for comparison
- "Group" contains the name of samples, which will be stored as colnames
- "numOtus" contains just the number of taxa that was found from the sample

File: R/loadFromMothur.R
Comment (TuomasBorman) [reply R/loadFromMothur.R]: x is sample names. I have named the variable differently now

File: R/loadFromMothur.R
Comment (TuomasBorman) [reply R/loadFromMothur.R]: So, should the `rownames(tse) `be set somehow?  I assumed that it is checked that `rownames(assay) == rownames(rowData)`. Otherwise, I think it throws an error when TSE is created: `"rowdata and assay are not parallel" `or something

File: R/loadFromMothur.R
Comment (TuomasBorman) [reply R/loadFromMothur.R]: rownames should match, and subsetting is unnecessary. I checked my data, and for some reason in taxonomic file random taxa are dropped from the middle

File: R/loadFromMothur.R
Comment (FelixErnst) [reply R/loadFromMothur.R]: All three variables should end up in `colData`, since you cannot expect all users to not need it.

Comment (TuomasBorman) [reply R/loadFromMothur.R]: Ok

File: R/loadFromMothur.R
Comment (FelixErnst) [reply R/loadFromMothur.R]: Having read the mothur man page, make sure to use their nomenclature. From the names you are using right now, it is for example not clear, in which file the count data is expected to be.

File: R/loadFromMothur.R
Comment (FelixErnst) [reply R/loadFromMothur.R]: Exactly. So it is critical, that you have a consistant strategy. If I remember correctly, assay names take precedence, when creating a `SE`.

Currently, you are doing the following steps:

- get assay data
- set the colnames of the `SE` of the assay as rownames, but then transpose the matrix
- get the rowData and set the rownames of the `SE` as rownames of the rowData
- get colData and just keep the columns which match the rownames, ... ah sorry the colnames, of the assay

So the strategy works for sure, but is not really consistent, since you are using three different types of using the dimnames

Good practice would be:

- to extract the assay at a single point
- to extract the colnames at a single point
- to extract the rowData at a single point
- to extract the rownames at a single point
- to extract the colData at a single point
- to combine the data and set the dimnames in a single step

Especially the subsetting/ordering of the colData leaves me with a weird feeling, since in the worst case scenario, `NA` get produced without the user getting notified, that the data is not matching. The second thing is the assay and dimname setting. I would do something along the lines of for the assay extraction:

```r
MOTHUR_NON_ASSAY_COLS <- c("label","numOtus","Group")
data <- read.table(featureTableFile, check.names=FALSE, header=TRUE,
                      sep="\t", stringsAsFactors=FALSE)
if(){
... # test for colnames not NULL
... # test for colnames present
}
assay <- as.matrix(data[,!(colnames(data) %in% MOTHUR_NON_ASSAY_COLS )])
assay <- t(assay)
colData <- DataFrame(data[,MOTHUR_NON_ASSAY_COLS])
return(list(assay = assay,
         colData = colData))
```

This way you don't touch any data, you don't have to and it is really easy to change, since you only have to change `MOTHUR_NON_ASSAY_COLS` to affect the outcome.

Comment (FelixErnst) [reply R/loadFromMothur.R]: For the rowData extraction, I would do something similar:

```r
 # Reads the file
MOTHUR_TAX_COL <- "Taxonomy"
data <- read.table(taxonomyTableFile, check.names=FALSE,
                      header=TRUE, sep="\t", stringsAsFactors=FALSE)
if(){
... # test for colnames present
... # check for MOTHUR_TAX_COL  present in colnames
}
# split taxa level into seperate columns
data [,MOTHUR_TAX_COL] <- gsub("[\"]", "", data [,MOTHUR_TAX_COL])
data [,MOTHUR_TAX_COL] <- gsub("[(1-100)]", "", data [,MOTHUR_TAX_COL])
tax <- tidyr::separate(data , MOTHUR_TAX_COL, into=c("Kingdom", "Phylum", "Order", "Class", "Family", "Genus"), sep=";", extra="merge")
tax$Genus <- gsub(";", "", tax$Genus)

# 
rowData <- cbind(DataFrame(tax),
                DataFrame(data[, !(colnames(data) %in% MOTHUR_TAX_COL)]))
if(!is.null(rowData$OTU)){
    if(){
    ... # check if the rownames are the same as in assay.
    }
    rownames(rowData) <- rowData$OTU
}
return(rowData)
```

Here again, you don't touch anything you don't need to and you make it clear, from the code which information takes precedence. If I review this code (I know, not the best example), then it is immediately clear, that the assay data rownames have the higher priority and the integrity of the data is ensured by the code.

File: R/loadFromMothur.R
Comment (FelixErnst) [reply R/loadFromMothur.R]: They should match, but if someone mixes up their files? A lot of `NA` get produced or a subset out of bound error is triggered, without any clear indication, what the problem is.

If the rownames should match, than you have to make sure, e.g. with this
```r
if(!all(.. == ...)){
   stop()
}
```

File: R/loadFromMothur.R
Comment (FelixErnst) [inline R/loadFromMothur.R]: As a discussion: The name `featureTableFile` is that used in the context of mothur? Is it a default name for mothur? I know, that the functions is informed a lot by `loadFromQIIME2`, but I think it not required to match the name.

I would put more emphasis on the view point of a typical mothur user. He has to recognize the file type by its name and know immediately, which file from mothur belong to which argument for `loadFromMothur`.

Is that the case for the argument names used here? I never have used mothur, so maybe someone else can offer an opinion? @antagomir @microsud

File: R/loadFromMothur.R
Comment (FelixErnst) [inline R/loadFromMothur.R]: Since the input from mothur, cannot require `TSE` to be used (now reference sequence, no tree), I would recommend just returning a `SE`.

File: R/loadFromMothur.R
Comment (TuomasBorman) [reply R/loadFromMothur.R]: I agree that it could be wise those change these names to match with Mothur

File: R/loadFromMothur.R
Comment (TuomasBorman) [reply R/loadFromMothur.R]: Changed

File: R/loadFromMothur.R
Comment (TuomasBorman) [reply R/loadFromMothur.R]: E.g.

assay = shared
rowData = constaxonomy
colData = design

However, I'm not sure if those are the names that are used in reality

Comment (FelixErnst) [reply R/loadFromMothur.R]: What do they use in their manual? If I remember correctly the `shared` name is also a documentation title, isn't it?

`taxonomyFile` and `designFile` are good names IMO. For the `sharedFile` a detailed entry on the man page might help as well.

Comment (TuomasBorman) [reply R/loadFromMothur.R]: Yes, those names that I listed are the names that they use as documentation titles

-    https://mothur.org/wiki/shared_file/
-    https://mothur.org/wiki/constaxonomy_file/
-    https://mothur.org/wiki/design_file/

Comment (FelixErnst) [reply R/loadFromMothur.R]: `constaxonomy` looks like a typo, but, hey, not our cup of coffee. `taxonomyFile` would probably match close enough, wouldn't it?

Comment (TuomasBorman) [reply R/loadFromMothur.R]: It might, but they actually have two different taxonomy files:

- taxonomy file https://mothur.org/wiki/taxonomy_file/
- constaxonomy file https://mothur.org/wiki/constaxonomy_file/

Now only `constaxonomy `is allowed. Should `taxonomy `file be allowed aswell? I was thinking, if the name is `taxonomyFile,` then it could be misleading, because in reality it is `constaxonomy `file

Comment (FelixErnst) [reply R/loadFromMothur.R]: Good catch. I think you could extend the current function to support both.
 I would implement like this:
- read in file without header
- check whether the first lines contains valid columns names
- If yes proceed as `constaxonomy`
- If no:
  - check that two columns exist
  - process the first column as rowData column `sequenceName`
  - process the second column into taxonomic groups and check the result
  - check that the dimension of the taxonomy matched the assay

I think that should yield a valid result.

File: R/loadFromMothur.R
Comment (TuomasBorman) [reply R/loadFromMothur.R]: tse --> se

File: DESCRIPTION
Comment (FelixErnst) [inline DESCRIPTION:77]: Why? Thats covered via `BiocStyle` isn't it?

Comment (TuomasBorman) [reply DESCRIPTION:77]: There was an error that said that it should be added there

Comment (TuomasBorman) [reply DESCRIPTION:77]: image

Comment (FelixErnst) [reply DESCRIPTION:77]: Good catch, but "Oh damn". I hope this doesn't crash a large number of the CRAN/Bioconductor builds. I haven't received any mail regarding this from the maintainer of knitr. So lets hope I don't have any other packages to modify.

File: tests/testthat/test-4IO.R
Comment (FelixErnst) [inline tests/testthat/test-4IO.R:66]: Can you add a few tests on the dimension? This ist just to make sure, that the `SE` actually conforms to the expectations (An `SE` can in theory have 0 rows or 0 columns).

Please check the following
- dimensions
- assays present
- rowData present + correct dimension
- colData present + correct dimension

Thanks

File: R/loadFromMothur.R
Comment (FelixErnst) [inline R/loadFromMothur.R]: These checks only work based on the name of the file. I think this is not really a solid solution, since problems can occur in to modes:
- wrong file name, right data
- right file name, wrong data

Mode 1 is not a severe problem, but mode 2 will produce unexpexted results. 

I would suggest the following:
- read in the data via `read.table`, because this is the common format (all files are tab-delimited files)
- check, if column names match certain assumptions. From what I can tell, this is actually the only requirement for correct processing of the data in following steps: A `shared` file requires the columns `"label","numOtus","Group"`, a taxonomy file either has the `Taxonomy` column or exactly `two` columns and the same number of rows than the shared file has data and a meta data file has column names and the same number of rows as the shared file has rows

Regarding the check implementation, it would be better to formulate the check differently. Instead of having a function return a type and do the check outside of the function, have a function like `.is_mothur_taxonomy_file` return a `boolean` result. That way the logic is contained in a single function and not separated into to two, the function and the check via `if`. This makes the code better maintainable and splits the logic for the individual files up. Each can be adjusted independently

File: tests/testthat/test-4IO.R
Comment (TuomasBorman) [reply tests/testthat/test-4IO.R:66]: Added tests

File: R/loadFromMothur.R
Comment (TuomasBorman) [reply R/loadFromMothur.R]: I modified those file type checks

**Review by FelixErnst — APPROVED** [review]
made a few formatting changes. Good to go after the checks clear. Thanks Tuomas for sticking with it! @tvborm

**Pull request merged**
